cli/src/pcluster/models/imagebuilder.py

```diff
@@ -185,24 +207,23 @@ def bucket(self):
             service_name=self.image_name,
             stack_name=self.image_name,
             custom_s3_bucket=custom_bucket_name,
-            artifact_directory=self.s3_artifacts_dir,
+            artifact_directory=self.s3_artifact_dir,
         )
         return self.__bucket
 
     def _get_custom_bucket(self):
         """Try to get custom bucket name from image tag or stack tag."""
         custom_bucket_name = None
         try:
-            image_info = AWSApi.instance().ec2.describe_image_by_name_tag(self.image_name)
-            custom_bucket_name = search_tag(image_info, PCLUSTER_S3_BUCKET_TAG)
-        except AWSClientError as e:
-            if not isinstance(e, ImageNotFoundError):
+            custom_bucket_name = self.image.s3_bucket_name
+        except ImageBuilderActionError as e:
+            if f"Image {self.image_name} does not exist." not in str(e):
                 raise ImageBuilderActionError(f"Unable to get S3 bucket name from image {self.image_name} tag. {e}")
 
         if custom_bucket_name is None:
             try:
                 custom_bucket_name = self.stack.s3_bucket_name
-            except AWSClientError as e:
+            except ImageBuilderActionError as e:
```

Comment:
Same here, you can define a NonExistingStackError.

The basic idea is that the model should raise an ImageBuilderAction to the "api" layer for all the known issues and this should be the only known exception to be catched by the api layer.

But for internal errors/exception we can define new exceptions and manage them in a different way.

Reply:
Added StackError and subclass NonExistingStackError to imagebuilder.py. For all the places calling stack property, we will catch StackError and then check if it is instance of NonExistingStackError or not. Therefore, calling stack property won't raise ImageBuilderActionError, instead raise ImageBuilderActionError in the caller.